qiskit/pulse/reschedule.py

```diff
@@ -180,19 +180,19 @@ def pad(schedule: Schedule, channels: Optional[Iterable[Channel]] = None,
         Schedule: The padded schedule
     """
     until = until or schedule.duration
-
     channels = channels or schedule.channels
-    occupied_channels = schedule.channels
-
-    unoccupied_channels = set(channels) - set(occupied_channels)
-
-    empty_timeslot_collection = schedule.timeslots.complement(until)
 
     for channel in channels:
-        for timeslot in empty_timeslot_collection.ch_timeslots(channel):
-            schedule |= Delay(timeslot.duration)(timeslot.channel).shift(timeslot.start)
-
-    for channel in unoccupied_channels:
-        schedule |= Delay(until)(channel)
+        if channel not in schedule.timeslots:
+            schedule |= Delay(until)(channel)
+            continue
+
+        curr_time = 0
+        for interval in schedule.timeslots[channel]:
+            if interval.start != curr_time:
```

Comment:
Just because `curr_time` is initialized to zero. We check if the next instruction starts in the future, if so, add a delay starting at `curr_time` up to `interval.start`. After that, `curr_time` is updated to the end of that next instruction, and repeat.

Looks like the existing code had a bug, which my code also has: `until` is not respected for a channel with timeslots going beyond `until`

Reply:
Ah, I thought `interval.start > curr_time` but it's equivalent with `interval.start != curr_time` because `interval.start >= curr_time == previous_interval.end` is implicitly guranteed.